proposals/0000-simplify-subsumption.md

```diff
@@ -0,0 +1,209 @@
+```
+g (\x y. f x y)
+```
+and now, of course, it is not surprising that the expression evaluates to 0.
+
+### Deep instantiation
+
+Suppose `f :: Int -> forall a. a -> a`.   Again, notice the forall to the right of the arrow.  Now consider this definition, which lacks a type signature:
+```
+g x = f
+```
+What type would you expect to infer for `g`?   The obvious answer (and the one we'd get without deep instantiation) is
+```
+g :: forall b. b -> Int -> forall a. a -> a
+```
+But GHC actually deeply instantiates `f` (for no very well-explained reason), so we get
+```
+g :: forall b a. b -> Int -> a -> a
+```
+with the buried foralls pulled to the top.  Perhaps that type is a tiny bit more explicable to the programmer.  But again, to produce that type, GHC must elaborate `g` to
+```
+g = /\ b a. \(x:b). \(i:Int). f i @a
+```
+GHC has eta-expaned `f`, which changes the semantics.  Yikes.
+
+### Contravariance
+
+Suppose you have
+```
+g :: ((forall a. a -> a) -> R) -> S
+f :: (Int -> Int) -> R
+```
+Now, is `(g f)` well typed?   That depends on whether
+```
+(Int -> Int) -> R   <=    (forall a. a -> a) -> R
+```
+where `<=` is pronounced “is more polymorphic than”  see the paper sections 4.4. and 4.6.
+
+Well, according to rule FUN of Figure 7, using contravariance of `(->)`, that is true if
+```
+forall a. a -> a    <=      Int -> Int
+```
+and that is certainly true.  But again, to witness that proof GHC needs to eta-expand during elaboration.  We get this elaboration of `(g f)`:
+```
+g (\(h : forall a. a->a).  f (h @Int))
+```
+Again we have changed the semantics.  Yikes.
+
+Again, lacking covariance the program would be rejected, but is easily fixed by manual eta-expansion, thus `g (\h -> f h)`
+
+### Covariance
+
+Fig 7 in the paper also supports covariance of the function arrow, but exactly the same eta-expansion issues arise.
+
+#
+
+
+## Proposed Change Specification
+
+There are no syntactic changes.
+
+The changes to the type system is to simplify the subsumption judgement
+by removing
+
+* Covariance of function types
+* Contravariance of function types
+* Deep skolemisation
+* Deep instantiation
+
+This could be under control of a flag, but I propose that we
+take a clear path to removing these features altogether, i.e.
+ultimately, without a flag that enables them.
+
+## Examples
+
+See Motivation above.
+
+## Effect and Interactions
+
+* See Section 7 of the [Quick Look paper](https://www.microsoft.com/en-us/research/publication/a-quick-look-at-impredicativity/) for a detailed analysis of the practical impact of these changes.
+* Everything (specification, implementation) becomes a bit simpler
+* Quick Look Impredicativity gains more power
+
+## Costs and Drawbacks
+
+The main user-facing cost is that some existing programs will require some manual eta-expansion.
+Jurriaan is gathering data on how many libraries are affected.
```

Comment:
This last remark is no longer accurate, and is superseded by the link to the paper above.